src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -2114,46 +2301,87 @@ private static BigInteger Subtract(uint[]? leftBits, int leftSign, uint[]? right
             left.AssertValid();
             right.AssertValid();
 
-            bool trivialLeft = left._bits == null;
-            bool trivialRight = right._bits == null;
+            return Multiply(left._bits, left._sign, right._bits, right._sign);
+        }
+
+        private static BigInteger Multiply(ReadOnlySpan<uint> left, int leftSign, ReadOnlySpan<uint> right, int rightSign)
+        {
+            bool trivialLeft = left.IsEmpty;
+            bool trivialRight = right.IsEmpty;
 
             if (trivialLeft && trivialRight)
             {
-                return (long)left._sign * right._sign;
+                return (long)leftSign * rightSign;
             }
 
+            BigInteger result;
+            uint[]? bitsFromPool = null;
+
             if (trivialLeft)
             {
-                Debug.Assert(right._bits != null);
-                uint[] bits = BigIntegerCalculator.Multiply(right._bits, NumericsHelpers.Abs(left._sign));
-                return new BigInteger(bits, bits, (left._sign < 0) ^ (right._sign < 0));
-            }
+                Debug.Assert(!right.IsEmpty);
 
-            if (trivialRight)
-            {
-                Debug.Assert(left._bits != null);
-                uint[] bits = BigIntegerCalculator.Multiply(left._bits, NumericsHelpers.Abs(right._sign));
-                return new BigInteger(bits, bits, (left._sign < 0) ^ (right._sign < 0));
+                int size = right.Length + 1;
+                Span<uint> bits = ((uint)size <= BigIntegerCalculator.StackAllocThreshold ?
+                                  stackalloc uint[BigIntegerCalculator.StackAllocThreshold]
+                                  : bitsFromPool = ArrayPool<uint>.Shared.Rent(size)).Slice(0, size);
+
+                BigIntegerCalculator.Multiply(right, NumericsHelpers.Abs(leftSign), bits);
+                result = new BigInteger(bits, (leftSign < 0) ^ (rightSign < 0));
             }
+            else if (trivialRight)
+            {
+                Debug.Assert(!left.IsEmpty);
 
-            Debug.Assert(left._bits != null && right._bits != null);
+                int size = left.Length + 1;
+                Span<uint> bits = ((uint)size <= BigIntegerCalculator.StackAllocThreshold ?
+                                  stackalloc uint[BigIntegerCalculator.StackAllocThreshold]
+                                  : bitsFromPool = ArrayPool<uint>.Shared.Rent(size)).Slice(0, size);
 
-            if (left._bits == right._bits)
-            {
-                uint[] bits = BigIntegerCalculator.Square(left._bits);
-                return new BigInteger(bits, bits, (left._sign < 0) ^ (right._sign < 0));
+                BigIntegerCalculator.Multiply(left, NumericsHelpers.Abs(rightSign), bits);
+                result = new BigInteger(bits, (leftSign < 0) ^ (rightSign < 0));
             }
+            else if (left.Length == right.Length && Unsafe.AreSame(ref Unsafe.AsRef(in left[0]), ref Unsafe.AsRef(in right[0])))
```

Comment:
This entire line should be simply `else if (left == right)`. The equality operator on `Span<T>` and `ReadOnlySpan<T>` is a shallow equality operator: it is contracted to check length and referential equality.

Reply:
Fixed